packages/cli/src/chat/chat-service.ts

```diff
@@ -0,0 +1,346 @@
+import { type RawData, WebSocket } from 'ws';
+import { z } from 'zod';
+
+import { ChatExecutionManager } from './chat-execution-manager';
+import {
+	chatMessageSchema,
+	type ChatMessage,
+	type ChatRequest,
+	Session,
+} from './chat-service.types';
+import {
+	getLastNodeExecuted,
+	getMessage,
+	isResponseNodeMode,
+	prepareMessageFromLastNode,
+	shouldResumeImmediately,
+} from './utils';
+
+const CHECK_FOR_RESPONSE_INTERVAL = 3000;
+const DRAIN_TIMEOUT = 50;
+const HEARTBEAT_INTERVAL = 30 * 1000;
+const HEARTBEAT_TIMEOUT = 60 * 1000;
+
+/**
+ * let frontend know that no user input is expected
+ */
+const N8N_CONTINUE = 'n8n|continue';
+/**
+ * send message for heartbeat check
+ */
+const N8N_HEARTBEAT = 'n8n|heartbeat';
+/**
+ * frontend did acknowledge the heartbeat
+ */
+const N8N_HEARTBEAT_ACK = 'n8n|heartbeat-ack';
+
+function closeConnection(ws: WebSocket) {
+	if (ws.readyState !== WebSocket.OPEN) return;
+
+	ws.once('drain', () => {
+		ws.close();
+	});
+
+	setTimeout(() => {
+		if (ws.readyState === WebSocket.OPEN) {
+			ws.close();
+		}
+	}, DRAIN_TIMEOUT);
+}
+
+@Service()
+export class ChatService {
+	private readonly sessions = new Map<string, Session>();
+	private heartbeatIntervalId: NodeJS.Timeout;
+
+	constructor(
+		private readonly executionManager: ChatExecutionManager,
+		private readonly logger: Logger,
+	) {
+		this.heartbeatIntervalId = setInterval(
+			async () => await this.checkHeartbeats(),
+			HEARTBEAT_INTERVAL,
+		);
+	}
+
+	async startSession(req: ChatRequest) {
+		const {
+			ws,
+			query: { sessionId, executionId, isPublic },
+		} = req;
+
+		if (!ws) {
+			throw new UnexpectedError('WebSocket connection is missing');
+		}
+
+		const execution = await this.executionManager.checkExecutionExists(executionId);
+
+		if (!execution) {
+			ws.send(`Execution with id "${executionId}" does not exist`);
+			ws.close(1008);
+			return;
+		}
+
+		ws.isAlive = true;
+
+		const key = `${sessionId}|${executionId}|${isPublic ? 'public' : 'integrated'}`;
+
+		if (this.sessions.has(key)) {
+			this.sessions.get(key)?.connection.terminate();
+			clearInterval(this.sessions.get(key)?.intervalId);
+			this.sessions.delete(key);
+		}
+
+		const onMessage = this.incomingMessageHandler(key);
+		const respondToChat = this.outgoingMessageHandler(key);
+
+		const intervalId = setInterval(async () => await respondToChat(), CHECK_FOR_RESPONSE_INTERVAL);
+
+		ws.once('close', () => {
+			ws.off('message', onMessage);
+			clearInterval(intervalId);
+			this.sessions.delete(key);
+		});
+
+		ws.on('message', onMessage);
+
+		const session: Session = {
+			connection: ws,
+			executionId,
+			sessionId,
+			intervalId,
+			isPublic: isPublic ?? false,
+			isProcessing: false,
+			lastHeartbeat: Date.now(),
+		};
+
+		this.sessions.set(key, session);
+
+		ws.send(N8N_HEARTBEAT);
+	}
+
+	private outgoingMessageHandler(sessionKey: string) {
+		return async () => {
+			let session: Session | undefined;
```

Comment:
This function is too complex as reported by eslint. Let's split it to not create more debt for ourselves. E.g. have own processing function for each of the main cases in the function so the main logic is easy to read